src/Strings.f90

```diff
@@ -364,81 +371,166 @@ FUNCTION cchar_to_fchar(cchar) RESULT(fchar)
-      ELSE
-        ! The first character was the delimiter so skip past it
-        stt = stt + LEN(separator)
-      ENDIF
-    ENDDO
-    ! If the string ends with a word, then it wasn't counted before
-    IF(.NOT.(this%s(stp-LEN(separator)+1:stp) == separator)) THEN
+    DO WHILE(sepLoc > 0)
       nSplits = nSplits + 1
-    ENDIF
+      stt = stt + sepLoc + LEN(separator) - 1
+      sepLoc = INDEX(this%s(stt:stp),separator)
+    ENDDO
+    !Account for strings that don't end in a delimiter
+    IF(stt <= stp) nSplits = nSplits + 1
     ALLOCATE(sub_str(nSplits))
 
-    ! Split along delimiters and store in the provided array
     stt = 1
     iSplit = 0
-    DO WHILE(INDEX(this%s(stt:stp),separator) > 0)
-      IF(INDEX(this%s(stt:stp),separator) > 1) THEN
-        iSplit = iSplit + 1
-        ! Strip out the string...subtract 2 (1 for exclusive, 1 for delimiter)
-        sub_str(iSplit) = &
-            this%s(stt:stt+INDEX(this%s(stt:stp),separator) - 2)
-        stt = stt + INDEX(this%s(stt:stp),separator) + LEN(separator) - 1
-      ELSE
-        stt = stt + LEN(separator)
-      ENDIF
+    sepLoc = INDEX(this%s(stt:stp),separator)
+    DO WHILE(sepLoc > 0)
+      iSplit = iSplit + 1
+      sub_str(iSplit) = this%s(stt:stt+sepLoc-2)
+      stt = stt + sepLoc + LEN(separator) - 1
+      sepLoc = INDEX(this%s(stt:stp),separator)
     ENDDO
     ! If the string ends with a word, then it wasn't snatched out before
-    IF(.NOT.(this%s(stp-LEN(separator)+1:stp) == separator)) THEN
+    IF(iSplit < nSplits) THEN
       sub_str(nSplits) = &
           this%s(INDEX(this%s(1:stp),separator,.TRUE.)+LEN(separator):stp)
     ENDIF
   ENDFUNCTION split_string
 !
 !-------------------------------------------------------------------------------
+!> @brief returns a copy of the string with every occurrence of a specified
+!> search pattern with the designated replacement pattern. This routine will
+!> automatically account for size changes of the underlying character array. If
+!> the search pattern is not found, then the original string is returned
+!>
+FUNCTION replace_pattern(this,oldPat,newPat) RESULT(retStr)
+  CLASS(StringType),INTENT(IN) :: this
+  CHARACTER(LEN=*),INTENT(IN) :: oldPat
+  CHARACTER(LEN=*),INTENT(IN) :: newPat
+  CHARACTER(LEN=:),ALLOCATABLE :: retStr
+  !
+  TYPE(StringType),ALLOCATABLE :: tokens(:)
+  INTEGER(SIK) :: iToken
+
+  !Split out the parts to be retained
+  tokens = this%split(oldPat)
+
+  !Set-up the new string
+  IF(SIZE(tokens) > 1) THEN
+    retStr = ''
+  ELSEIF(SIZE(tokens) == 1) THEN
+    retStr = char(tokens(1))
+    RETURN
+  ELSE
+    retStr = ''
+    RETURN
+  ENDIF
+
+  DO iToken=1,SIZE(tokens)-1
+    IF(tokens(iToken) == '') THEN
+      retStr = retStr//newPat
+    ELSE
+      retStr = retStr//tokens(iToken)//newPat
+    ENDIF
+  ENDDO
+  IF(tokens(SIZE(tokens)) /= '') THEN
+    retStr = retStr//tokens(SIZE(tokens))
+  ELSE
+    retStr = retStr//newPat
+  ENDIF
```

Comment:
The related question first...in this routine yes since it specifies the delimiter in the split routine it invokes the algorithm that will return empty strings.

I have to look at this and its test. it seems backward. There should be two cases. One the original string ends in a delimiter and should be appended with the new pattern. The other it ends in a part of the string that remains and that should be appended to the end.

Reply:
stupid `not equal`. 

So back to the original question, yes that can happen again. `str%split(' ')` would split "A B C D  " (NOTE: two spaces after 'D') into `(/"A","B","C","D",""/)` because it is everything to the left of the delimiter just found. Then each _token_ requires a new pattern to be appended except the last one, which only gets the new pattern if it itself was the delimiter.

Reply:
Ha, ok, so the `==`/`/=` was backwards... that helps clarify things a bit.  I hadn't quite worked through it enough to know that was the issue, but it seemed off.

Ok, I didn't know when I wrote this comment that `split()` and `split(' ')` would have different behaviors.  I think it makes sense now (with the fixed equality sign, I think)